gcloud-java-core/src/main/java/com/google/cloud/GrpcServiceOptions.java

```diff
@@ -0,0 +1,229 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud;
+
+import com.google.cloud.spi.ServiceRpcFactory;
+import com.google.common.base.Preconditions;
+
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.internal.SharedResourceHolder.Resource;
+
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Abstract class representing service options for those services that use gRPC as the transport
+ * layer.
+ *
+ * @param <ServiceT> the service subclass
+ * @param <ServiceRpcT> the spi-layer class corresponding to the service
+ * @param <OptionsT> the {@code ServiceOptions} subclass corresponding to the service
+ */
+public abstract class GrpcServiceOptions<ServiceT extends Service<OptionsT>, ServiceRpcT,
+    OptionsT extends GrpcServiceOptions<ServiceT, ServiceRpcT, OptionsT>>
+    extends ServiceOptions<ServiceT, ServiceRpcT, OptionsT> {
+
+  private static final long serialVersionUID = 6415982522610509549L;
+  private final int initialTimeout;
+  private final double timeoutMultiplier;
+  private final int maxTimeout;
+
+  /**
+   * Shared thread pool executor.
+   */
+  private static final Resource<ScheduledExecutorService> EXECUTOR =
+      new Resource<ScheduledExecutorService>() {
+        @Override
+        public ScheduledExecutorService create() {
+          ScheduledThreadPoolExecutor service = new ScheduledThreadPoolExecutor(8);
+          service.setKeepAliveTime(5, TimeUnit.SECONDS);
+          service.allowCoreThreadTimeOut(true);
+          service.setRemoveOnCancelPolicy(true);
+          return service;
+        }
+
+        @Override
+        public void close(ScheduledExecutorService instance) {
+          instance.shutdown();
+        }
+      };
+
+  /**
+   * An interface that provides access to a scheduled executor service.
+   */
+  public interface ExecutorProvider {
+
+    /**
+     * Returns the scheduled executor service.
+     */
+    ScheduledExecutorService get();
+
+    /**
+     * Shuts down the scheduled executor service if it is no longer used.
+     */
+    void shutdown();
+  }
+
+  /**
+   * An interface that provides access to a scheduled executor service.
+   */
+  private static class DefaultExecutorProvider implements ExecutorProvider {
+
+    private ScheduledExecutorService service;
+    private boolean shutdown = false;
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid using redundant field initializer for 'shutdown'](https://www.codacy.com/app/mziccard/gcloud-java/file/2575054385/issues/source?bid=3123109&fileBranchId=3299216#l89)